{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Firstly, sequences in fasta files were renamed, because the names were too long. The renamed fasta was aligned to the genome, turned to bed and gtf produced out of the bed file was used. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"/home/anna/laba/merged_lncRNAs_transcripts_filtered_nonredundant.out_30_1_plus\", \"r\") as infa:\n",
    "    with open(\"/home/anna/laba/merged_lncRNAs_transcripts_filtered_nonredundant_new.fasta\", \"w\") as outfa:\n",
    "        with open(\"/home/anna/laba/merged_lncRNAs_transcripts_filtered_nonredundant_tocutfasta.csv\", \"w\") as pot:\n",
    "            i = 0\n",
    "            for line in infa:\n",
    "                if line.startswith(\">\"):\n",
    "                    i += 1\n",
    "                    outfa.write(\">lnc_tr\")\n",
    "                    outfa.write(str(i) + \"\\n\")\n",
    "                    pot.write(line.strip() + \";\" + \">lnc_tr\" + str(i) + \"\\n\") # table of new and old names matching \n",
    "                else:\n",
    "                    outfa.write(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SCANTI2 needs reads, gtf annotation and a reference. Gtf annotation was produced from fasta file of lncRNAs with **fqtobed.sh** script but using **bedtools bamtobed** with **-bed12** option. Then \n",
    "#### ./bedToGenePred lnc_tr_12.bed stdout | ./genePredToGtf file stdin lnc_tr_12.gtf\n",
    "to produce gtf.\n",
    "\n",
    "As we used only reads mapped to lncRNAs, the reads extracted with the following commands: \n",
    "#### bedtools intersect  -s -b /home/andrey/anna/lnc_200.bed -abam ~/anna/gams_filt2_sort.bam > lnc200_gam.bam\n",
    "#### samtools fastq lnc200_gam.bam > lnc200_gam.fastq\n",
    "\n",
    "For SQANTI2 read names should be in a special format. This script converts read names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open( \"/home/andrey/anna/lnc200_gam.fastq\", \"r\") as infa:\n",
    "    with open( \"/media/ilia/anna/lnc200_gam_renamed.fastq\", \"w\") as outfa:\n",
    "        with open(\"/media/ilia/anna/rename_matrix_lnc200_gam.csv\", \"w\") as pot:\n",
    "            i = -1\n",
    "            for line in infa:\n",
    "                if line.startswith(\"@\"):\n",
    "                    i += 1\n",
    "                    output = \"@PB.\" + str(i) + \".0\"\n",
    "                    outfa.write(output + \"\\n\")\n",
    "                    pot.write(line.strip() + \";\" + output + \"\\n\")\n",
    "                else:\n",
    "                    outfa.write(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SQANTI2 commands used:\n",
    "\n",
    "#### python sqanti_qc2.py --skipORF  --aligner_choice minimap2 -o lnc_tr_gam /media/ilia/anna/lnc200_gam_renamed2.fastq ~/anna/lnc_tr_12.gtf ~/minimap2/Ppatens_318_v3.fa 2> ls-error.log &\n",
    "#### python sqanti_filter2.py lnc_tr_gam_classification.txt lnc_tr_gam_corrected.fasta lnc_tr_gam_corrected.gtf &\n",
    "\n",
    "Results were processed in python."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>exons</th>\n",
       "      <th>structural_category</th>\n",
       "      <th>associated_gene</th>\n",
       "      <th>associated_transcript</th>\n",
       "      <th>ref_length</th>\n",
       "      <th>ref_exons</th>\n",
       "      <th>...</th>\n",
       "      <th>CDS_genomic_end</th>\n",
       "      <th>predicted_NMD</th>\n",
       "      <th>perc_A_downstream_TTS</th>\n",
       "      <th>seq_A_downstream_TTS</th>\n",
       "      <th>dist_to_cage_peak</th>\n",
       "      <th>within_cage_peak</th>\n",
       "      <th>dist_to_polya_site</th>\n",
       "      <th>within_polya_site</th>\n",
       "      <th>polyA_motif</th>\n",
       "      <th>polyA_dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PB.0.0</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>-</td>\n",
       "      <td>731</td>\n",
       "      <td>2</td>\n",
       "      <td>full-splice_match</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lnc_tr14390</td>\n",
       "      <td>705.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>GCAAGTGAGCCCCTTCCTGT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PB.1.0</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>-</td>\n",
       "      <td>852</td>\n",
       "      <td>2</td>\n",
       "      <td>full-splice_match</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lnc_tr14390</td>\n",
       "      <td>705.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>CAAGCAAGTGAGCCCCTTCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PB.10.0</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>-</td>\n",
       "      <td>466</td>\n",
       "      <td>2</td>\n",
       "      <td>full-splice_match</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lnc_tr14390</td>\n",
       "      <td>705.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>ACTGTATATCTGTTCGAGAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PB.100.0</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>+</td>\n",
       "      <td>529</td>\n",
       "      <td>3</td>\n",
       "      <td>full-splice_match</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lnc_tr4414</td>\n",
       "      <td>830.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>TTAGTGGTTTTCAAACTATT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PB.1000.0</td>\n",
       "      <td>Chr01</td>\n",
       "      <td>+</td>\n",
       "      <td>988</td>\n",
       "      <td>1</td>\n",
       "      <td>full-splice_match</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lnc_tr4829</td>\n",
       "      <td>201.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>TTTTGCTCCACACTTGAGAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     isoform  chrom strand  length  exons structural_category associated_gene  \\\n",
       "0     PB.0.0  Chr01      -     731      2   full-splice_match             NaN   \n",
       "1     PB.1.0  Chr01      -     852      2   full-splice_match             NaN   \n",
       "2    PB.10.0  Chr01      -     466      2   full-splice_match             NaN   \n",
       "3   PB.100.0  Chr01      +     529      3   full-splice_match             NaN   \n",
       "4  PB.1000.0  Chr01      +     988      1   full-splice_match             NaN   \n",
       "\n",
       "  associated_transcript  ref_length  ref_exons  ...  CDS_genomic_end  \\\n",
       "0           lnc_tr14390       705.0        2.0  ...              NaN   \n",
       "1           lnc_tr14390       705.0        2.0  ...              NaN   \n",
       "2           lnc_tr14390       705.0        2.0  ...              NaN   \n",
       "3            lnc_tr4414       830.0        3.0  ...              NaN   \n",
       "4            lnc_tr4829       201.0        1.0  ...              NaN   \n",
       "\n",
       "   predicted_NMD  perc_A_downstream_TTS  seq_A_downstream_TTS  \\\n",
       "0            NaN                   15.0  GCAAGTGAGCCCCTTCCTGT   \n",
       "1            NaN                   25.0  CAAGCAAGTGAGCCCCTTCC   \n",
       "2            NaN                   25.0  ACTGTATATCTGTTCGAGAC   \n",
       "3            NaN                   25.0  TTAGTGGTTTTCAAACTATT   \n",
       "4            NaN                   25.0  TTTTGCTCCACACTTGAGAA   \n",
       "\n",
       "  dist_to_cage_peak  within_cage_peak dist_to_polya_site  within_polya_site  \\\n",
       "0               NaN               NaN                NaN                NaN   \n",
       "1               NaN               NaN                NaN                NaN   \n",
       "2               NaN               NaN                NaN                NaN   \n",
       "3               NaN               NaN                NaN                NaN   \n",
       "4               NaN               NaN                NaN                NaN   \n",
       "\n",
       "   polyA_motif  polyA_dist  \n",
       "0          NaN         NaN  \n",
       "1          NaN         NaN  \n",
       "2          NaN         NaN  \n",
       "3          NaN         NaN  \n",
       "4          NaN         NaN  \n",
       "\n",
       "[5 rows x 45 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full = pd.read_csv('lnc_tr_gam_classification.filtered_lite_classification.txt', sep='\\t')\n",
    "full.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is a large table where a row is a single read. Many columns have many NA or are empty at all. Remove them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "well = full[['isoform', 'chrom', 'strand', 'length', 'exons', 'structural_category', \"associated_gene\", 'associated_transcript', 'ref_length',  \"ref_exons\", \"subcategory\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Thus we can get only fully proved transcripts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoform</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>length</th>\n",
       "      <th>exons</th>\n",
       "      <th>structural_category</th>\n",
       "      <th>associated_gene</th>\n",
       "      <th>ref_length</th>\n",
       "      <th>ref_exons</th>\n",
       "      <th>subcategory</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>associated_transcript</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>lnc_tr1</th>\n",
       "      <td>PB.118996.0</td>\n",
       "      <td>Chr22</td>\n",
       "      <td>+</td>\n",
       "      <td>573</td>\n",
       "      <td>1</td>\n",
       "      <td>full-splice_match</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1374.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lnc_tr100</th>\n",
       "      <td>PB.57090.0</td>\n",
       "      <td>Chr06</td>\n",
       "      <td>-</td>\n",
       "      <td>185</td>\n",
       "      <td>1</td>\n",
       "      <td>full-splice_match</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3262.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lnc_tr1000</th>\n",
       "      <td>PB.14374.0</td>\n",
       "      <td>Chr04</td>\n",
       "      <td>-</td>\n",
       "      <td>648</td>\n",
       "      <td>1</td>\n",
       "      <td>full-splice_match</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1240.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lnc_tr10029</th>\n",
       "      <td>PB.10497.0</td>\n",
       "      <td>Chr02</td>\n",
       "      <td>-</td>\n",
       "      <td>179</td>\n",
       "      <td>1</td>\n",
       "      <td>full-splice_match</td>\n",
       "      <td>NaN</td>\n",
       "      <td>596.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mono-exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lnc_tr10036</th>\n",
       "      <td>PB.10866.0</td>\n",
       "      <td>Chr02</td>\n",
       "      <td>-</td>\n",
       "      <td>501</td>\n",
       "      <td>6</td>\n",
       "      <td>full-splice_match</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2638.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>multi-exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           isoform  chrom strand  length  exons  \\\n",
       "associated_transcript                                             \n",
       "lnc_tr1                PB.118996.0  Chr22      +     573      1   \n",
       "lnc_tr100               PB.57090.0  Chr06      -     185      1   \n",
       "lnc_tr1000              PB.14374.0  Chr04      -     648      1   \n",
       "lnc_tr10029             PB.10497.0  Chr02      -     179      1   \n",
       "lnc_tr10036             PB.10866.0  Chr02      -     501      6   \n",
       "\n",
       "                      structural_category  associated_gene  ref_length  \\\n",
       "associated_transcript                                                    \n",
       "lnc_tr1                 full-splice_match              NaN      1374.0   \n",
       "lnc_tr100               full-splice_match              NaN      3262.0   \n",
       "lnc_tr1000              full-splice_match              NaN      1240.0   \n",
       "lnc_tr10029             full-splice_match              NaN       596.0   \n",
       "lnc_tr10036             full-splice_match              NaN      2638.0   \n",
       "\n",
       "                       ref_exons subcategory  \n",
       "associated_transcript                         \n",
       "lnc_tr1                      1.0   mono-exon  \n",
       "lnc_tr100                    1.0   mono-exon  \n",
       "lnc_tr1000                   1.0   mono-exon  \n",
       "lnc_tr10029                  1.0   mono-exon  \n",
       "lnc_tr10036                  6.0  multi-exon  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proved = well[well.structural_category == 'full-splice_match']\n",
    "proved.groupby(proved.associated_transcript).min().head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
